### PR TITLE
Quant op dev

### PR DESCRIPTION
**Pull request by huangxu96**
### PR types

New features
### PR changes

OPs
### Describe

Implemented  the ChannelWiseQuantizeDequantizeAbsMaxOp which is used for quant aware training. The input is a tensor 
(fp32), the output (fp32) is the a tensor be channel-wised quantized and dequantized sequentially. The quantized scales of each channel will also be output.

image

### CODE REVIEW

File: paddle/fluid/operators/fake_quantize_op.cc [hunk 1]
```diff
@@ -291,6 +321,62 @@ In above three formulas, the range value of c is as follow:
...
+
+  void InferShape(framework::InferShapeContext* ctx) const override {
+    OP_INOUT_CHECK(ctx->HasInput("X"), "Input", "X",
+                   "FakeChannelWiseQuantizeDequantizeAbsMax");
+    OP_INOUT_CHECK(ctx->HasOutput("Out"), "Output", "Out",
+                   "FakeChannelWiseQuantizeDequantizeAbsMax");
+    OP_INOUT_CHECK(ctx->HasOutput("OutScale"), "Output", "OutScale",
+                   "FakeChannelWiseQuantizeDequantizeAbsMax");
+    ctx->SetOutputDim("Out", ctx->GetInputDim("X"));
+    ctx->SetOutputDim("OutScale", {ctx->GetInputDim("X")[0]});
```
Comment (wzzju) [inline paddle/fluid/operators/fake_quantize_op.cc]: Suggestion: Add an attribute in `FakeChannelWiseQuantizeDequantizeAbsMaxOpMaker` named `quant_dim` or other. Here, not use 0, but use `ctx->Attrs().Get<int>("quant_dim")`.

File: paddle/fluid/operators/fake_quantize_op.cc [hunk 2]
```diff
@@ -291,6 +321,62 @@ In above three formulas, the range value of c is as follow:
...
+                                "the received is %d",
+                                bit_length));
+        });
+    AddComment(R"DOC(
+The scale of FakeChannelWiseQuantize operator is a vector.
+In detail, each channel of the input X has a scale value.
+
+$$scale_c = max(abs(X_c))$$
+$$range = 2^{bit\_length - 1} - 1$$
+$$Out_c = round(\frac{X_c * range} {scale_c})$$
```
Comment (wzzju) [inline paddle/fluid/operators/fake_quantize_op.cc]: This is just the process of the quantization, not including dequantization.

$$Out_c = round(\frac{X_c * range} {scale_c}) * \frac{scale_c} {range}$$

File: paddle/fluid/operators/fake_quantize_op.cu [hunk 3]
```diff
@@ -325,8 +325,49 @@ struct FindMovingAverageAbsMaxFunctor<platform::CUDADeviceContext, T> {
...
+    const T* in_data = in.data<T>();
+    const T* scale_data = scale.data<T>();
+    T* out_data = out->mutable_data<T>(ctx.GetPlace());
+
+    ChannelClipAndQuantDequantKernel<T><<<grid, block, 0, ctx.stream()>>>(
+        in_data, scale_data, bin_cnt, num, channel, out_data);
+  }
+};
+
+template struct ChannelClipFakeQuantDequantFunctor<platform::CUDADeviceContext,
```
Comment (wzzju) [inline paddle/fluid/operators/fake_quantize_op.cu:502]: Please think about this:
* What does this statement do?
* Is this statement necessary？

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py [hunk 4]
```diff
@@ -248,10 +251,11 @@ def __init__(self,
         # For FakeQuant
         self._fake_quant_weight = _get_fake_quant_type(
             weight_quantize_type, self.weight.name, moving_rate, weight_bits,
-            self._dtype, True)
+            self._dtype, True, self.weight.shape[0])
         self._fake_quant_input = _get_fake_quant_type(
             activation_quantize_type,
-            layer.full_name(), moving_rate, activation_bits, self._dtype, False)
+            layer.full_name(), moving_rate, activation_bits, self._dtype, False,
+            self.weight.shape[0])
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: Use `self.weight` is not necessary here, you may give `channel_num` a default value in `_get_fake_quant_type`.

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py [hunk 5]
```diff
@@ -210,12 +210,15 @@ def forward(self, input):
 
 
 def _get_fake_quant_type(quant_type, name, moving_rate, quant_bits, dtype,
-                         quant_on_weight):
+                         quant_on_weight, channel_num):
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]:
```python
def _get_fake_quant_type(quant_type, name, moving_rate, quant_bits, dtype,
                                            quant_on_weight=False, channel_num=None):

if quant_type == 'channel_wise_abs_max':
    assert channel_num is not None, "some info..."

```

File: python/paddle/fluid/contrib/slim/quantization/imperative/qat.py [hunk 6]
```diff
@@ -98,7 +98,9 @@ def __init__(self,
         self._activation_bits = activation_bits
         self._moving_rate = moving_rate
 
-        quant_type = {'abs_max', 'moving_average_abs_max'}
+        quant_type = {
+            'abs_max', 'moving_average_abs_max', 'channel_wise_abs_max'
+        }
         if activation_quantize_type not in quant_type:
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/qat.py:108]: Please add a statement described  as below:

```python
        assert activation_quantize_type != 'channel_wise_abs_max', \
            "The activation quantization type does not support 'channel_wise_abs_max'."
```

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py [hunk 7]
```diff
@@ -318,6 +322,8 @@ def __init__(self,
                  moving_rate=0.9,
                  weight_quantize_type='abs_max',
                  activation_quantize_type='abs_max'):
+        if weight_quantize_type == 'channel_wise_abs_max':
+            weight_quantize_type = 'abs_max'
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: Add some `TODO`s: To support channel-wise quantization on Linear.

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py [hunk 8]
```diff
@@ -327,10 +333,11 @@ def __init__(self,
         # For FakeQuant
         self._fake_quant_weight = _get_fake_quant_type(
             weight_quantize_type, self.weight.name, moving_rate, weight_bits,
-            self._dtype, True)
+            self._dtype, True, self.weight.shape[0])
         self._fake_quant_input = _get_fake_quant_type(
             activation_quantize_type,
-            layer.full_name(), moving_rate, activation_bits, self._dtype, False)
+            layer.full_name(), moving_rate, activation_bits, self._dtype, False,
+            self.weight.shape[0])
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: The same issue as above.

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py [hunk 9]
```diff
@@ -373,3 +380,82 @@ def forward(self, input):
...
+            attrs = ('bit_length', self._quant_bits)
+            quant_out = _varbase_creator(
+                type=input.type,
+                name="{}channelwise.quantized.dequantized".format(input.name),
+                shape=input.shape,
+                dtype=input.dtype,
+                persistable=False)
+
+            out_scale = self._scale
+            if not out_scale[0]:
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: Error here, please correct it as follow:
```python
if not out_scale:
```

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py [hunk 10]
```diff
@@ -1701,7 +1701,7 @@ def apply(self, graph):
                         quant_var_node = dequantized_vars_map[arg_name]
                     else:
                         quant_var_node, _ = \
-                            self._inser_quant_dequant_moving_average_abs_max_op(
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]: Don't modify this file, which is irrelevant to this PR.

File: python/paddle/fluid/contrib/slim/tests/test_imperative_qat.py [hunk 11]
```diff
@@ -157,7 +157,8 @@ class TestImperativeQat(unittest.TestCase):
 
     def test_qat_save(self):
         imperative_qat = ImperativeQuantAware(
-            weight_quantize_type='abs_max',
+            #weight_quantize_type='abs_max',
+            weight_quantize_type='channel_wise_abs_max',
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/tests/test_imperative_qat.py]: Don't modify this file, please add your own unit test.

File: python/paddle/fluid/contrib/slim/tests/test_imperative_qat.py [hunk 12]
```diff
@@ -275,7 +275,8 @@ def _build_static_lenet(main, startup, is_test=False, seed=1000):
 
         reader = paddle.batch(
             paddle.dataset.mnist.test(), batch_size=32, drop_last=True)
-        weight_quantize_type = 'abs_max'
+        #weight_quantize_type = 'abs_max'
+        weight_quantize_type = 'channel_wise_abs_max'
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/tests/test_imperative_qat.py]: Don't modify this file, please add your own unit test.

**Review by wzzju — CHANGES_REQUESTED** [review]

File: python/paddle/fluid/contrib/slim/tests/test_imperative_qat.py [hunk 13]
```diff
@@ -157,7 +157,8 @@ class TestImperativeQat(unittest.TestCase):
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/tests/test_imperative_qat.py]: Please add `test_qat_save` unit  test for `channel_wise_abs_max `.

File: paddle/fluid/operators/fake_quantize_op.cc [hunk 14]
```diff
@@ -291,6 +321,62 @@ In above three formulas, the range value of c is as follow:
...
+
+class FakeChannelWiseQuantizeDequantizeAbsMaxOpMaker
+    : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() override {
+    AddInput("X", "(Tensor) Input is float data type.");
+    AddOutput("Out",
+              "(Tensor) Output of quantized and dequantized low level tensor, "
+              "saved as float data type.");
+    AddOutput("OutScale", "(Tensor) Current channel wise scale");
```
Comment (juncaipeng) [inline paddle/fluid/operators/fake_quantize_op.cc:454]: Please refer to #25639, add attribute of  `quant_axis`. 
For conv2d and depthwise_conv2d, the dims of weights is Cout * Cin * H * W, so the quant axis is 0. 
For mul and conv2d_tranpose, the dims of weights if Cin * Cout * (H * W), the quant axis  is 1.

File: paddle/fluid/operators/fake_quantize_op.cc (hunk 2 above)
Comment (huangxu96) [reply paddle/fluid/operators/fake_quantize_op.cc]: Fixed

File: paddle/fluid/operators/fake_quantize_op.cc (hunk 14 above)
Comment (huangxu96) [reply paddle/fluid/operators/fake_quantize_op.cc:454]: Thanks! Already referred your code and learnt a lot.

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py (hunk 4 above)
Comment (huangxu96) [reply python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: Fixed

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py (hunk 5 above)
Comment (huangxu96) [reply python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: Fixed

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py (hunk 8 above)
Comment (huangxu96) [reply python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: Fixed

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py (hunk 10 above)
Comment (huangxu96) [reply python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]: Made it as before.

File: python/paddle/fluid/contrib/slim/tests/test_imperative_qat.py (hunk 11 above)
Comment (huangxu96) [reply python/paddle/fluid/contrib/slim/tests/test_imperative_qat.py]: Cool, added one new test.

File: python/paddle/fluid/contrib/slim/tests/test_imperative_qat.py (hunk 12 above)
Comment (huangxu96) [reply python/paddle/fluid/contrib/slim/tests/test_imperative_qat.py]: Fixed

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py [hunk 15]
```diff
@@ -209,13 +211,107 @@ def forward(self, input):
...
+        else:
+            self._scale = None
+
+    def forward(self, input):
+        if in_dygraph_mode():
+            attrs = ('bit_length', self._quant_bits, 'quant_axis',
+                     self._quant_axis)
+            quant_out = _varbase_creator(
+                type=input.type,
+                name="{}channelwise.quantized.dequantized".format(input.name),
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: Please use `name="{}.quantized.dequantized".format(input.name)`.

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py [hunk 16]
```diff
@@ -209,13 +211,107 @@ def forward(self, input):
...
+                 dtype='float32',
+                 quant_on_weight=False):
+        super(FakeChannelWiseQuantDequantAbsMax, self).__init__()
+        self._quant_bits = quant_bits
+        self._quant_axis = quant_axis
+        self._dtype = dtype
+        self._name = name
+        self._channel_num = channel_num
+        scale_prefix = "{}.scale".format(
+            name) if name else 'channel_wise_quant_dequant.scale'
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: Please use ` scale_prefix = "{}.scale".format(name) if name else 'quant_dequant.scale'`.

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py [hunk 17]
```diff
@@ -209,13 +211,107 @@ def forward(self, input):
...
+
+
+def _get_fake_quant_type(quant_type,
+                         name,
+                         moving_rate,
+                         quant_bits,
+                         dtype,
+                         quant_on_weight=False,
+                         channel_num=None,
+                         quant_axis=0):
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: Use  the follow code instead.
```python
def _get_fake_quant_type(quant_type, **kwargs):
```

Then you should use keyword arguments to call the `_get_fake_quant_type` function.

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py [hunk 18]
```diff
@@ -209,13 +211,107 @@ def forward(self, input):
...
+        assert channel_num is not None, "You need input channel_num when you using channel_wise_abs_max strategy."
+
     fake_quant_map = {
         'abs_max':
         lambda: FakeQuantAbsMax(name, quant_bits, dtype, quant_on_weight),
         'moving_average_abs_max':
-        lambda: FakeQuantMovingAverage(name, moving_rate, quant_bits, dtype)
+        lambda: FakeQuantMovingAverage(name, moving_rate, quant_bits, dtype),
+        'channel_wise_abs_max':
+        lambda: FakeChannelWiseQuantDequantAbsMax(name, channel_num, quant_bits, quant_axis, dtype, quant_on_weight)
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: Please use the follow code instead.
```python
    fake_quant_map = {
        'abs_max':
        lambda: FakeQuantAbsMax(kwargs["name"], kwargs["quant_bits"], kwargs["dtype"], kwargs["quant_on_weight"]),
        'moving_average_abs_max':
        lambda: FakeQuantMovingAverage(kwargs["name"], kwargs["moving_rate"], kwargs["quant_bits"], kwargs["dtype"]),
        'channel_wise_abs_max':
        lambda: FakeChannelWiseQuantDequantAbsMax(kwargs["name"], kwargs["channel_num"], kwargs["quant_bits"], kwargs["quant_axis"], kwargs["dtype"], kwargs["quant_on_weight"])
    }
```

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py [hunk 19]
```diff
@@ -248,7 +344,7 @@ def __init__(self,
         # For FakeQuant
         self._fake_quant_weight = _get_fake_quant_type(
             weight_quantize_type, self.weight.name, moving_rate, weight_bits,
-            self._dtype, True)
+            self._dtype, True, self.weight.shape[0])
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: Please use keyword arguments to call the `_get_fake_quant_type` function.

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py [hunk 20]
```diff
@@ -327,7 +423,8 @@ def __init__(self,
         # For FakeQuant
         self._fake_quant_weight = _get_fake_quant_type(
             weight_quantize_type, self.weight.name, moving_rate, weight_bits,
-            self._dtype, True)
+            self._dtype, True, self.weight.shape[LINEAR_QUANT_AXIS],
+            LINEAR_QUANT_AXIS)
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: * Please use keyword arguments to call the `_get_fake_quant_type` function.
* Don't use the global var `LINEAR_QUANT_AXIS `, and you can use a member var instead.

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py [hunk 21]
```diff
@@ -24,9 +24,11 @@
 
 __all__ = [
     'FakeQuantMovingAverage', 'FakeQuantAbsMax', 'QuantizedConv2D',
-    'QuantizedLinear'
+    'QuantizedLinear', 'FakeChannelWiseQuantDequantAbsMax'
 ]
 
+LINEAR_QUANT_AXIS = 1
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: Don't use the global var `LINEAR_QUANT_AXIS `.

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py [hunk 22]
```diff
@@ -209,13 +211,107 @@ def forward(self, input):
...
-def _get_fake_quant_type(quant_type, name, moving_rate, quant_bits, dtype,
-                         quant_on_weight):
+class FakeChannelWiseQuantDequantAbsMax(layers.Layer):
+    def __init__(self,
+                 name=None,
+                 channel_num=None,
+                 quant_bits=8,
+                 quant_axis=0,
+                 dtype='float32',
+                 quant_on_weight=False):
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py:219]: assert `quant_on_weight ` == False ?

**Review by wzzju — CHANGES_REQUESTED** [review]

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py (hunk 15 above)
Comment (huangxu96) [reply python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: Done

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py (hunk 16 above)
Comment (huangxu96) [reply python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: Done

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py (hunk 20 above)
Comment (huangxu96) [reply python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: Done

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py (hunk 21 above)
Comment (huangxu96) [reply python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: Done

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py (hunk 19 above)
Comment (huangxu96) [reply python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: Done

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py (hunk 18 above)
Comment (huangxu96) [reply python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: Done

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py (hunk 17 above)
Comment (huangxu96) [reply python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: Done

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py (hunk 22 above)
Comment (huangxu96) [reply python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py:219]: I think it should be assert quant_on_weight == True.

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py [hunk 23]
```diff
@@ -209,6 +209,89 @@ def forward(self, input):
...
+        else:
+            self._scale = None
+
+    def forward(self, input):
+        if in_dygraph_mode():
+            attrs = ('bit_length', self._quant_bits, 'quant_axis',
+                     self._quant_axis)
+            quant_out = _varbase_creator(
+                type=input.type,
+                name="{}quantized.dequantized".format(input.name),
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: Please use `name="{}.quantized.dequantized".format(input.name)`.

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py [hunk 24]
```diff
@@ -341,13 +429,16 @@ def __init__(self,
         self.weight = getattr(layer, 'weight')
         self.bias = getattr(layer, 'bias')
         # For FakeQuant
+        self.linear_quant_axis = 1
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: Please use `self._linear_quant_axis = 1` instead.

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py [hunk 25]
```diff
@@ -261,7 +348,8 @@ def __init__(self,
             moving_rate=moving_rate,
             quant_bits=weight_bits,
             dtype=self._dtype,
-            quant_on_weight=True)
+            quant_on_weight=True,
+            channel_num=self.weight.shape[0])
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: Please add the following code:
```python
self._conv2d_quant_axis = 0
self._fake_quant_weight = _get_fake_quant_type(
        weight_quantize_type,
        name=self.weight.name,
        moving_rate=moving_rate,
        quant_bits=weight_bits,
        dtype=self._dtype,
        quant_on_weight=True,
        channel_num=self.weight.shape[self._conv2d_quant_axis],
        quant_axis=self._conv2d_quant_axis)
```

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py (hunk 22 above)
Comment (wzzju) [reply python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py:219]: Yes, you are right.

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py (hunk 24 above)
Comment (huangxu96) [reply python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: Done

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py (hunk 23 above)
Comment (huangxu96) [reply python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: Done

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py (hunk 25 above)
Comment (huangxu96) [reply python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: Done

**Review by wzzju — APPROVED** [review]
LGTM.

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py (hunk 9 above)
Comment (huangxu96) [reply python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: Done

**Review by zhiqiu — APPROVED** [review]
LGTM for op_function_generator.cc

**Review by wzzju — APPROVED** [review]
LGTM.

**Pull request merged**